### PR TITLE
feat: Create Image Element

### PR DESCRIPTION
**Pull request by dskamiotis**
co-authored with @SHession 

## What does this change?
This PR adds an Image Element replicating the existing Embed element in Composer. However this element includes more functionality and buttons to access the Grid for images.

This PR also:
- adds a `CustomDropDown` which renders a modified Source Dropdown (allowing for clearer displayed text and improved - arrow button).
- includes the use of `createTextField` as a text area allowing for multi line 
- adds a createTextField with multiline capabilities set with a default of 4 lines.
- Field labels included as prop from Component
- Accesses Grid instantly when image element button selected
- Buttons to re crop/change selected image and copy caption to alt text field
- Logic to deal with Assets from Grid defaulting to those of 1000px width and the next largest as a fallback
- Adds styling for width of image to stretch to 100% of column
- Styling includes Source library from buttons and columns
- Adds field data on initial insert image element press
- Update fields on selecting image for source, photographer and caption
- Fixes demo image element with demoSetMedia, decoupling the demo and actual image element
- setMedia object used to house data from Grid and refactor for use with the element, as a MediaPayload type

### Images
| Selecting image | 
Screenshot 2021-08-06 at 16 05 00 

| Caption button |
Screenshot 2021-08-06 at 15 55 00 

| Fields input and validation |
Screenshot 2021-08-06 at 15 55 00 

| Re-crop button |
Screenshot 2021-08-06 at 15 55 00 

## How to test
Run `yarn start`
Passes all integration and unit test

### CODE REVIEW

**Review by SHession — COMMENTED** [review]
I've noticed when we "copy from caption" in ProseMirror elements, we keep any marks included i.e. bold or italics. These should be stripped out, as with the Composer image element. I think we can specify no allowed marks in the text field nodeSpec, unless something relies on this.

EDIT: PR (#90) should resolve this issue.

**Review by SHession — COMMENTED** [review]
The wording on the error messages are a little different to Composer and it would be nice if we could make them consistent: 

Composer:
image

PME: 
image

The PME text field will also expand over multiple lines, Composer will not.


EDIT: The wording isn't really an issue, but the multilines may be more so.

**Review by SHession — COMMENTED** [review]
The "Copy from caption" button in Composer is placed next to the "Alt text" label, this makes sense as it suggests it will update with "Alt Text" with the "Caption" text. It would be good to be consistent with Composer in this regard, currently PME places the button above the label.

Composer:
image

PME:
image

**Review by SHession — COMMENTED** [review]
It's possible to add a small crop which causes Composer complain, it might be worth investigating and replicating this behaviour.

Composer: 
image

**Review by SHession — COMMENTED** [review]
We are currently adding a default menu-bar in PME which is inconsistent with the rest of Composer. As the Composer menu-bar lives in Composer it may be tricky to include it the project. This might be one for the integrating this element into Composer card. 

PME: 
image

File: src/elements/image/ImageElement.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      {
+        mediaId: undefined,
+        mediaApiUri: undefined,
+        assets: [],
+        suppliersReference: "",
+      },
+      { openImageSelector }
+    ),
+    source: createTextField(),
+    weighting: createCustomField("none-selected", [
```
Comment (SHession) [inline src/elements/image/ImageElement.tsx:76]: I believe weighting should be called "role"
image

**Review by rhystmills — APPROVED** [review]
Looking really good. I agree with Sam's point about the position of the Alt Text button, and made a couple of minor comments, but overall it's looking good 👍 

Tests all passing too.

I've realised that the Source `Button` and `Select` components use _Guardian Text Sans_ rather than _Guardian Agate Sans_ (as we do elsewhere in Composer). This might be worth changing in general but it's by no means essential to do in this PR.

**Comment by dskamiotis** [comment]
Thanks @rhystmills really good spot, well worth keeping consistent font styling. I may make this as a small change to this PR as opposed to a new one.

File: demo/helpers.ts [hunk 2]
```diff
@@ -55,15 +105,43 @@ export const onSelectImage = (setMedia: SetMedia) => {
...
 
   (document.querySelector(
     ".modal__body iframe"
-  ) as HTMLIFrameElement).src = `https://media.test.dev-gutools.co.uk/images/${mediaId}`;
+  ) as HTMLIFrameElement).src = mediaId
+    ? `https://media.test.dev-gutools.co.uk/images/${mediaId}`
+    : `https://media.test.dev-gutools.co.uk/`;
 
-  const listener = onGridMessage(setMedia, modal);
+  modal.style.display = "Inherit";
```
[the remark this answers is not in the document]
Comment (dskamiotis) [reply demo/helpers.ts:143]: 👍

File: demo/helpers.ts [hunk 3]
```diff
@@ -55,15 +105,43 @@ export const onSelectImage = (setMedia: SetMedia) => {
...
+
+  window.addEventListener("message", listener, {
+    once: true,
+  });
+
+  document.querySelector(".modal__dismiss")?.addEventListener(
+    "click",
+    () => {
+      window.removeEventListener("message", listener);
+      modal.style.display = "None";
```
[the remark this answers is not in the document]
Comment (dskamiotis) [reply demo/helpers.ts:128]: 👍

File: src/elements/image/ImageElement.tsx (hunk 1 above)
Comment (dskamiotis) [reply src/elements/image/ImageElement.tsx:76]: @rhystmills in Composer it seems to be Weighting, unless something has changed?

File: src/elements/demo-image/DemoImageElement.tsx [hunk 4]
```diff
@@ -31,13 +31,13 @@ type ImageField = {
...
-  onSelectImage: (setMedia: SetMedia) => void;
-  onCropImage: (mediaId: string, setMedia: SetMedia) => void;
+  onSelectImage: (setMedia: DemoSetMedia) => void;
+  onCropImage: (mediaId: string, setMedia: DemoSetMedia) => void;
 };
 
 export const createImageFields = (
-  onSelectImage: (setSrc: SetMedia) => void,
-  onCropImage: (mediaId: string, setMedia: SetMedia) => void
+  onSelectImage: (setSrc: DemoSetMedia) => void, // once selected image
```
[the remark this answers is not in the document]
Comment (dskamiotis) [reply src/elements/demo-image/DemoImageElement.tsx:39]: no this was me most probably, good spot!

**Comment by dskamiotis** [comment]
> Agate Sans

Interesting question, we could change the font from the source of `source` in the index.d.ts, and that would apply the font to all buttons used in future from within PME. This seems one solution?
Alternatively locally overriding the styling for the individual button?

**Pull request merged**